packages/dds/shared-object-base/src/sharedObject.ts

```diff
@@ -519,3 +447,89 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
 
     protected abstract applyStashedOp(content: any): unknown;
 }
+
+/**
+ * SharedObject with simplified, synchronous summarization and GC
+ */
+export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedObjectEvents>
+    extends SharedObjectCore<TEvent> {
+    /**
+     * True while we are garbage collecting this object's data.
+     */
+    private _isGCing: boolean = false;
+
+    protected get serializer(): IFluidSerializer {
+        /**
+         * During garbage collection, the SummarySerializer keeps track of IFluidHandles that are serialized. These
+         * handles represent references to other Fluid objects.
+         *
+         * This is fine for now. However, if we implement delay loading in DDss, they may load and de-serialize content
+         * in summarize. When that happens, they may incorrectly hit this assert and we will have to change this.
+         */
+        assert(!this._isGCing,
+            0x075 /* "SummarySerializer should be used for serializing data during summary." */);
+        return super.serializer;
+    }
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).getAttachSummary}
+     */
+    public getAttachSummary(fullTree: boolean = false, trackState: boolean = false): ISummaryTreeWithStats {
+        return this.summarizeCore(this.serializer, fullTree);
+    }
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).summarize}
+     */
+    public async summarize(fullTree: boolean = false, trackState: boolean = false): Promise<ISummaryTreeWithStats> {
+        return this.summarizeCore(this.serializer, fullTree);
+    }
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).getGCData}
+     */
+    public getGCData(fullGC: boolean = false): IGarbageCollectionData {
+        // Set _isGCing to true. This flag is used to ensure that we only use SummarySerializer to serialize handles
+        // in this object's data.
+        assert(!this._isGCing, 0x078 /* "Possible re-entrancy! Summary should not already be in progress." */);
+        this._isGCing = true;
+
+        let gcData: IGarbageCollectionData;
+        try {
+            const serializer = new SummarySerializer(
+                this.runtime.channelsRoutingContext,
+                (handle: IFluidHandle) => this.handleDecoded(handle),
+            );
+            gcData = this.getGCDataCore(serializer);
```

Comment:
I find it a bit weird the usage of serializer arg and return value. It feels like only one of them should be used, i.e.
1. either we pass serializer and expect it to be used, with this function interrogating it after the call to extract GC data.
2. OR we do not pass serializer, and let callers do whatever they want RE how they built up GC info.

Given that all getGCDataCore() implementations seems to do serializer.getSerializedRoutes() dance at the end, I'm thinking we want # 1 with whoever wanting more control using SharedObjectCore class and overwriting getGCData() method instead.